File: src/main/java/com/naminhyeok/fantazzk/room/domain/RoomTeamMember.java
package com.naminhyeok.fantazzk.room.domain;

import com.naminhyeok.fantazzk.room.RoomId;
import java.time.Instant;
import java.util.Objects;

public final class RoomTeamMember {
    private final RoomTeamMemberId roomTeamMemberId;
    private final RoomId roomId;
    private final String teamLeaderId;
    private final String playerName;
    private final int assignOrder;
    private final Instant createdAt;
    private final Instant updatedAt;

    private RoomTeamMember(
            RoomTeamMemberId roomTeamMemberId,
            RoomId roomId,
            String teamLeaderId,
            String playerName,
            int assignOrder,
            Instant createdAt,
            Instant updatedAt
    ) {
        this.roomTeamMemberId = roomTeamMemberId == null ? RoomTeamMemberId.random() : roomTeamMemberId;
        this.roomId = Objects.requireNonNull(roomId, "roomId");
        this.teamLeaderId = requireText(teamLeaderId, "팀장 식별자는 비어 있을 수 없습니다");
        this.playerName = requireText(playerName, "선수 이름은 비어 있을 수 없습니다");
        this.assignOrder = assignOrder;
        this.createdAt = Objects.requireNonNull(createdAt, "createdAt");
        this.updatedAt = Objects.requireNonNull(updatedAt, "updatedAt");
    }

    public static RoomTeamMember create(RoomId roomId, String teamLeaderId, String playerName, int assignOrder) {
        Instant now = Instant.now();
        return new RoomTeamMember(null, roomId, teamLeaderId, playerName, assignOrder, now, now);
    }

    public static RoomTeamMember restore(
            RoomTeamMemberId roomTeamMemberId,
            RoomId roomId,
            String teamLeaderId,
            String playerName,
            int assignOrder,
            Instant createdAt,
            Instant updatedAt
    ) {
        return new RoomTeamMember(roomTeamMemberId, roomId, teamLeaderId, playerName, assignOrder, createdAt, updatedAt);
    }

    public RoomTeamMember copy() {
        return restore(roomTeamMemberId, roomId, teamLeaderId, playerName, assignOrder, createdAt, updatedAt);
    }

    public RoomTeamMemberId getRoomTeamMemberId() {
        return roomTeamMemberId;
    }

    public RoomId getRoomId() {
        return roomId;
    }

    public String getTeamLeaderId() {
        return teamLeaderId;
    }

    public String getPlayerName() {
        return playerName;
    }

    public int getAssignOrder() {
        return assignOrder;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof RoomTeamMember roomTeamMember)) {
            return false;
        }
        return assignOrder == roomTeamMember.assignOrder
                && roomTeamMemberId.equals(roomTeamMember.roomTeamMemberId)
                && roomId.equals(roomTeamMember.roomId)
                && teamLeaderId.equals(roomTeamMember.teamLeaderId)
                && playerName.equals(roomTeamMember.playerName)
                && createdAt.equals(roomTeamMember.createdAt)
                && updatedAt.equals(roomTeamMember.updatedAt);
    }

    @Override
    public int hashCode() {
        return Objects.hash(roomTeamMemberId, roomId, teamLeaderId, playerName, assignOrder, createdAt, updatedAt);
    }

    private static String requireText(String value, String message) {
        String normalized = Objects.requireNonNull(value, "value").trim();
        if (normalized.isEmpty()) {
            throw new IllegalArgumentException(message);
        }
        return normalized;
    }
}

File: src/main/kotlin/com/naminhyeok/fantazzk/room/domain/RoomTeamMember.kt
package com.naminhyeok.fantazzk.room.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.jmolecules.ddd.types.Identifier
import java.time.Instant

@Entity
@Table(name = "room_team_member")
class RoomTeamMember protected constructor(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private var persistentId: Long? = null,
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "room_id", nullable = false)
    private var room: Room? = null,
    @Column(name = "team_leader_id", nullable = false)
    val teamLeaderId: String = "",
    @Column(name = "player_name", nullable = false)
    val playerName: String = "",
    @Column(name = "assign_order", nullable = false)
    val assignOrder: Int = 0,
    @Column(name = "created_at", nullable = false)
    val createdAt: Instant = Instant.now(),
    @Column(name = "updated_at", nullable = false)
    val updatedAt: Instant = Instant.now(),
) {
    internal constructor(
        roomTeamMemberId: Long? = null,
        roomId: Long?,
        teamLeaderId: String,
        playerName: String,
        assignOrder: Int,
        createdAt: Instant = Instant.now(),
        updatedAt: Instant = Instant.now(),
    ) : this(
        roomTeamMemberId = roomTeamMemberId?.takeIf { it != 0L }?.let(::RoomTeamMemberId),
        roomId = roomId?.takeIf { it != 0L }?.let(::RoomId),
        teamLeaderId = teamLeaderId,
        playerName = playerName,
        assignOrder = assignOrder,
        createdAt = createdAt,
        updatedAt = updatedAt,
    )

    constructor(
        roomTeamMemberId: RoomTeamMemberId? = null,
        roomId: RoomId? = null,
        teamLeaderId: String,
        playerName: String,
        assignOrder: Int,
        createdAt: Instant = Instant.now(),
        updatedAt: Instant = Instant.now(),
    ) : this(
        persistentId = roomTeamMemberId?.value,
        room = roomId?.let { Room.reference(it.value) },
        teamLeaderId = teamLeaderId,
        playerName = playerName,
        assignOrder = assignOrder,
        createdAt = createdAt,
        updatedAt = updatedAt,
    )

    internal constructor(
        roomId: Long? = null,
        teamLeaderId: String,
        playerName: String,
        assignOrder: Int,
        createdAt: Instant = Instant.now(),
        updatedAt: Instant = Instant.now(),
    ) : this(
        roomTeamMemberId = null,
        roomId = roomId,
        teamLeaderId = teamLeaderId,
        playerName = playerName,
        assignOrder = assignOrder,
        createdAt = createdAt,
        updatedAt = updatedAt,
    )

    val id: RoomTeamMemberId?
        get() = persistentId?.let(::RoomTeamMemberId)

    internal val roomTeamMemberId: Long
        get() = persistentId ?: 0L

    internal val roomId: Long
        get() = room?.roomId ?: 0L

    internal fun attach(room: Room): RoomTeamMember = apply { this.room = room }

    internal fun detachCopy(): RoomTeamMember =
        RoomTeamMember(
            roomTeamMemberId = id,
            roomId = room?.persistedIdOrNull(),
            teamLeaderId = teamLeaderId,
            playerName = playerName,
            assignOrder = assignOrder,
            createdAt = createdAt,
            updatedAt = updatedAt,
        )

    internal fun copy(
        roomTeamMemberId: Long = this.roomTeamMemberId,
        roomId: Long = this.roomId,
        teamLeaderId: String = this.teamLeaderId,
        playerName: String = this.playerName,
        assignOrder: Int = this.assignOrder,
        createdAt: Instant = this.createdAt,
        updatedAt: Instant = this.updatedAt,
    ): RoomTeamMember =
        RoomTeamMember(
            roomTeamMemberId = roomTeamMemberId.takeIf { it != 0L },
            roomId = roomId,
            teamLeaderId = teamLeaderId,
            playerName = playerName,
            assignOrder = assignOrder,
            createdAt = createdAt,
            updatedAt = updatedAt,
        )
}

data class RoomTeamMemberId(
    val value: Long,
) : Identifier {
    init {
        require(value > 0L) { "RoomTeamMemberId는 0보다 커야 합니다" }
    }
}
